protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs

```diff
@@ -0,0 +1,1075 @@
+        &mut self,
+        request_id: u32,
+        downstream_hash_rate: f32,
+        is_header_only: bool,
+        id: u32,
+    ) -> Result<Vec<Mining>, Error> {
+        match is_header_only {
+            true => {
+                self.new_standard_channel_for_hom_downstream(request_id, downstream_hash_rate, id)
+            }
+            false => self.new_standard_channel_for_non_hom_downstream(
+                request_id,
+                downstream_hash_rate,
+                id,
+            ),
+        }
+    }
+
+    pub fn new_extended_channel(
+        &mut self,
+        request_id: u32,
+        hash_rate: f32,
+        min_extranonce_size: u16,
+    ) -> Option<Vec<Mining>> {
+        let extended_channels_group = 0;
+        let max_extranonce_size = self.extranonces.get_range2_len() as u16;
+        if min_extranonce_size <= max_extranonce_size {
+            // SECURITY is very unlikely to finish the ids btw this unwrap could be used by an attaccher that
+            // want to dirsrupt the service maybe we should have a method to reuse ids that are no
+            // longer connected?
+            let channel_id = self
+                .ids
+                .safe_lock(|ids| ids.new_channel_id(extended_channels_group).unwrap())
+                .unwrap();
+            let target = crate::utils::hash_rate_to_target(hash_rate, self.share_per_min);
+            let extranonce = self
+                .extranonces
+                .next_extended(max_extranonce_size as usize)?;
+            let extranonce_prefix = extranonce.into_prefix(self.extranonces.get_prefix_len())?;
+            let success = OpenExtendedMiningChannelSuccess {
+                request_id,
+                channel_id,
+                target,
+                extranonce_size: max_extranonce_size,
+                extranonce_prefix,
+            };
+            self.extended_channels.insert(channel_id, success.clone());
+            let mut result = vec![Mining::OpenExtendedMiningChannelSuccess(success)];
+            if let Some((new_prev_hash, _)) = &self.last_prev_hash {
+                result.push(Mining::SetNewPrevHash(new_prev_hash.clone()))
+            };
+            if let Some((job, _)) = &self.last_valid_job {
+                result.push(Mining::NewExtendedMiningJob(job.clone()))
+            };
+            for (job, _) in &self.future_jobs {
+                result.push(Mining::NewExtendedMiningJob(job.clone()))
+            }
+            Some(result)
+        } else {
+            Some(vec![Mining::OpenMiningChannelError(
+                OpenMiningChannelError::new_max_target_out_of_range(request_id),
+            )])
+        }
+    }
+
+    pub fn on_new_prev_hash_from_tp(
+        &mut self,
+        m: &SetNewPrevHashFromTp<'static>,
+    ) -> Result<u32, Error> {
+        let job_id = self.job_creator.on_new_prev_hash(m).unwrap_or(0);
+        let new_prev_hash = SetNewPrevHash {
+            channel_id: 0,
```

Comment:
why is channel_id always 0?

Reply:
Good catch. So, here we set the channel to 0 cause this prev hash i valid for every channel created. I used 0 cause is kinda (internally) reserved cause channels start from 1.
A better solution could be use another type with only the PrevHash field that we need like:
```rust
struct ReducedPHash {
   job_id: u32,
   ...
}
```
then `ChannelFactory::new_prevhash` will be `Option<ReducedPHash>`
then `ChannelFactory::on_new_prev_hash` will take a `ReducedPHash`
then `ProxyExtendedChannelFactory::on_new_prev_hash` will transform the `SetNewPrevHash` in a `ReducedPHash` before calling `ChannelFactory::on_new_prev_hash` 

And when we need a `SetNewPrevHash` we use the right channel id. Right now you can set `SetNewPrevHash::channel_id` at whatever number you want and everything should keep working cause is not checked anywhere (not sure if it should be, not sure also if that field have any practical utility)

I will fix the code to use the ReducedPHash struct and set the right channel_id 

Same thing for `new_extended_job()` in `job_creators`  we are creating an extended job that will be used by all the channels so we put 0 as id I will modify the code in a way similar to the one that I proposed above also here.

Reply:
Your call. if you think we can just add comments to explain 0 is the general case, I am good with that too.

If you think the changing to a new type improves stability I have a couple questions:
1. would it make sense to create a separate issue for this?
2. would an `Option<u32>` be easier than a new type? Then the during serialization the `Some/None` checks can be enforced. 
3. If the new Type is the best way, what do you think about changing `Reduced` to `Staged` (or some other word that implies channel_id needs to be added to make a SNPH)? "Reduced" gives the connotation fields have been removed for different reasons.

Reply:
1. I would prefer do it in this PR
2. The SetNewPrevHash is a Sv2 message and channel_id is not optional make it optional would mean changing the spec so, using Option<u32> and changing the spec would harder then just internally  defining a new type.
3. Agree

* the new type is just used internally in the lib what is been sent is SetNewPrevHash with the right channel id (or group id when we send it to proxy that groups channels)